src/Orleans.Runtime/GrainDirectory/GrainLocator.cs

```diff
@@ -0,0 +1,120 @@
+        public GrainLocator(IGrainDirectory grainDirectory, InClusterGrainLocator inClusterGrainLocator)
+        {
+            this.grainDirectory = grainDirectory;
+            this.inClusterGrainLocator = inClusterGrainLocator;
+            this.cache = new LRUBasedGrainDirectoryCache(GrainDirectoryOptions.DEFAULT_CACHE_SIZE, GrainDirectoryOptions.DEFAULT_MAXIMUM_CACHE_TTL);
+        }
+
+        public async Task<List<ActivationAddress>> Lookup(GrainId grainId)
+        {
+            if (grainId.IsClient)
+                return await this.inClusterGrainLocator.Lookup(grainId);
+
+            var results = new List<ActivationAddress>();
+
+            var entry = await this.grainDirectory.Lookup(grainId.ToParsableString());
+
+            if (entry == null)
+                return results;
+
+            results.Add(ConvertToActivationAddress(entry));
+            this.cache.AddOrUpdate(grainId, results.Select(item => Tuple.Create(item.Silo, item.Activation)).ToList(), 0);
+
+            return results;
+        }
+
+        public async Task<ActivationAddress> Register(ActivationAddress address)
+        {
+            if (address.Grain.IsClient)
+                return await this.inClusterGrainLocator.Register(address);
+
+            var result = await this.grainDirectory.Register(ConvertToGrainAddress(address));
+            var activationAddress = ConvertToActivationAddress(result);
+            this.cache.AddOrUpdate(
+                activationAddress.Grain,
+                new List<Tuple<SiloAddress,ActivationId>>() { Tuple.Create(activationAddress.Silo, activationAddress.Activation) },
+                0);
+            return activationAddress;
+        }
+
+        public bool TryLocalLookup(GrainId grainId, out List<ActivationAddress> addresses)
+        {
+            if (grainId.IsClient)
+                return this.inClusterGrainLocator.TryLocalLookup(grainId, out addresses);
+
+            if (this.cache.LookUp(grainId, out var results))
+            {
+                addresses = results
+                    .Select(tuple => ActivationAddress.GetAddress(tuple.Item1, grainId, tuple.Item2))
+                    .ToList();
+                return true;
+            }
+
+            addresses = null;
+            return false;
+        }
+
+        public async Task Unregister(ActivationAddress address, UnregistrationCause cause)
+        {
+            if (address.Grain.IsClient)
+                await this.inClusterGrainLocator.Unregister(address, cause);
+
+            await this.grainDirectory.Unregister(ConvertToGrainAddress(address));
+            this.cache.Remove(address.Grain);
+        }
+
+        public async Task UnregisterMany(List<ActivationAddress> addresses, UnregistrationCause cause)
+        {
+            var tasks = addresses.Select(addr => Unregister(addr, cause)).ToList();
+            await Task.WhenAll(tasks);
+            foreach (var addr in addresses)
+                this.cache.Remove(addr.Grain);
```

Comment:
Clients are currently never cached, but they will be removed here. Also, this seems to remove grains from cache twice, since Unregister already calls cache.Remove

Reply:
Forgot a return here. Client are cached in the `LocalGrainDirectory` implementation